src/fusiontensor/base_interface.jl

```diff
@@ -0,0 +1,165 @@
+end
+
+# tensor contraction is a block data_matrix product.
+function Base.:*(left::FusionTensor, right::FusionTensor)
+  @assert matching_dual(domain_axes(left), codomain_axes(right))
+  new_data_matrix = data_matrix(left) * data_matrix(right)
+  return fusiontensor(new_data_matrix, codomain_axes(left), domain_axes(right))
+end
+
+Base.:+(ft::FusionTensor) = ft
+
+# tensor addition is a block data_matrix add.
+function Base.:+(left::FusionTensor, right::FusionTensor)
+  @assert matching_axes(axes(left), axes(right))
+  new_data_matrix = data_matrix(left) + data_matrix(right)
+  return FusionTensor(
+    new_data_matrix, codomain_axes(left), domain_axes(left), trees_block_mapping(left)
+  )
+end
+
+function Base.:-(ft::FusionTensor)
+  new_data_matrix = -data_matrix(ft)
+  return FusionTensor(
+    new_data_matrix, codomain_axes(ft), domain_axes(ft), trees_block_mapping(ft)
+  )
+end
+
+function Base.:-(left::FusionTensor, right::FusionTensor)
+  @assert matching_axes(axes(left), axes(right))
+  new_data_matrix = data_matrix(left) - data_matrix(right)
+  return FusionTensor(
+    new_data_matrix, codomain_axes(left), domain_axes(left), trees_block_mapping(left)
+  )
+end
+
+function Base.:/(ft::FusionTensor, x::Number)
+  return FusionTensor(
+    data_matrix(ft) / x, codomain_axes(ft), domain_axes(ft), trees_block_mapping(ft)
+  )
+end
+
+Base.Array(ft::FusionTensor) = Array(cast_to_array(ft))
+
+# adjoint is costless: dual axes, swap codomain and domain, take data_matrix adjoint.
+# data_matrix coeff are not modified (beyond complex conjugation)
+transpose_mapping(d::Dict) = Dict([reverse(k) => transpose_mapping(v) for (k, v) in d])
+function transpose_mapping(b::BlockIndexRange{2})
+  new_block = Block(reverse(Tuple(Block(b))))
+  return new_block[reverse(b.indices)...]
+end
+function Base.adjoint(ft::FusionTensor)
+  return FusionTensor(
+    adjoint(data_matrix(ft)),
+    dual.(domain_axes(ft)),
+    dual.(codomain_axes(ft)),
+    transpose_mapping(trees_block_mapping(ft)),
+  )
+end
+
+Base.axes(ft::FusionTensor) = (codomain_axes(ft)..., domain_axes(ft)...)
+
+# conj is defined as coefficient wise complex conjugation, without axis dual
+Base.conj(ft::FusionTensor{<:Real}) = ft   # same object for real element type
+function Base.conj(ft::FusionTensor)
+  return FusionTensor(
+    conj(data_matrix(ft)), codomain_axes(ft), domain_axes(ft), trees_block_mapping(ft)
+  )
+end
+
+function Base.copy(ft::FusionTensor)
+  return FusionTensor(
+    copy(data_matrix(ft)),
+    copy.(codomain_axes(ft)),
+    copy.(domain_axes(ft)),
+    copy(trees_block_mapping(ft)),
+  )
+end
+
+function Base.deepcopy(ft::FusionTensor)
+  return FusionTensor(
+    deepcopy(data_matrix(ft)),
+    deepcopy.(codomain_axes(ft)),
+    deepcopy.(domain_axes(ft)),
+    deepcopy(trees_block_mapping(ft)),
+  )
+end
+
+# eachindex is automatically defined for AbstractArray. We do not want it.
+Base.eachindex(::FusionTensor) = error("eachindex not defined for FusionTensor")
```

Comment:
Would the correct definition be:
```julia
function Base.eachindex(::FusionTensor)
  ax = blockedrange.(block_dimensions.((codomain_axes(ft)..., domain_axes(ft)...)))
  return CartesianIndices(ax)
end
```
? But also shouldn't we define:
```julia
function Base.axes(::FusionTensor)
  return blockedrange.(block_dimensions.((codomain_axes(ft)..., domain_axes(ft)...)))
end
```
? I.e. I would think these should be defined so that `axes(::FusionTensor)`, `size(::FusionTensor)`, `eachindex(::FusionTensor)` match what you would get for those functions if you cast the `FusionTensor` to an array, i.e. `axes(ft) == axes(to_array(ft))`.

Reply:
Something that is confusing about the current set of definitions is that `size(ft) != length.(axes(ft))`. So I think we may just need different kinds of size/axes definitions, i.e.:
```julia
fusiontensor_axes(ft::FusionTensor) = (codomain_axes(ft)..., domain_axes(ft)...)
fusiontensor_size(ft::FusionTensor) = length.(fusiontensor_axes(ft))

Base.axes(ft::FusionTensor) = blockedrange.(block_dimensions.(fusiontensor_axes(ft)))
Base.size(ft::FusionTensor) = length.(axes(ft))
```
Probably `fusiontensor_axes`/`fusiontensor_size` isn't the best naming convention but you get the idea.

Reply:
It seems like we need some terminology for this, like `reduced_axes`/`reduced_size`, `fused_axes`/`fused_size`, etc.

Reply:
An issue with the definition of `axes` I suggest above is that you don't get the property that `similar(a, axes(a))` gets you back an object of type `a`, i.e. it can't recreate that the object should be a `FusionTensor` from the axes.

Maybe there should be a special type:
```julia
struct FusionTensorAxes{N,Ncodomain,Ndomain,CodomainAxes<:Tuple{Vararg{AbstractUnitRange,Ncodomain}},DomainAxes<:Tuple{Vararg{AbstractUnitRange,Ndomain}}}
  codomain_axes::CodomainAxes
  domain_axes::DomainAxes
end
```
that is the output of `axes(::FusionTensor)`. That doesn't quite get us the property `size(ft::FusionTensor) == length.(axes(ft))` when there are non-abelian symmetries, somehow we want graded axes with non-abelian symmetries to have two kinds of lengths.

Reply:
So something like:
```julia
struct AbelianUnitRange{T,Range<:AbstractGradedUnitRange{T}} <: AbstractBlockedUnitRange{T}
  range::Range
end
Base.length(r::AbelianUnitRange) = sum(blocklengths(r))
BlockArrays.blocklengths(r::AbelianUnitRange) = block_dimensions(r)
```

Reply:
Right, I agree, but there are two issues I'm bringing up:
1. `axes(::FusionTensor)` doesn't output bipartitioned axes (as you say, I think `FusionTensorAxes` as I defined above can solve that issue if it acts like a flat collection of the axes but also has information about the bipartitioning).
2. The issue that:
```julia
using FusionTensors: FusionTensor
using GradedUnitRanges: dual, gradedrange
using SymmetrySectors: SU2
g = gradedrange([SU2(0) => 1, SU2(1//2) => 2])
ft = FusionTensor(Float64, (g,), (dual(g),))
@show length.(axes(ft))
@show size(ft)
```
outputs:
```julia
length.(axes(ft)) = (3, 3)
size(ft) = (5, 5)
```
right now because of how `length(::AbstractGradedUnitRange)` is defined.

My previous posts are probably a bit confusing but I'm trying to think through how to address both of these issues, my proposal for the second issue is basically that `FusionTensorAxes` could actually iterate over `AbelianUnitRange` wrapping the domain and codomain axes of the current `FusionTensor` definition so that `length` of each axis is defined in a way that is consistent with Base and the current definition of `size(::FusionTensor)`.

Reply:
Or we could change the definition of `BlockArrays.blocklengths(::AbstractGradedUnitRange)`/`Base.length(::AbstractGradedUnitRange)` for non-abelian symmetries, I'm not sure what the implications of that would be.

Reply:
I think I am definitely not following entirely. I think the current output should be considered a bug: `length(axes(ft, 1))` really should be `5`, there are 5 basis vectors in that vector space. There are only three independent parameters, but that seems to me like it should not influence the `length`?

Reply:
> I think I am definitely not following entirely. I think the current output should be considered a bug: `length(axes(ft, 1))` really should be `5`, there are 5 basis vectors in that vector space. There are only three independent parameters, but that seems to me like it should not influence the `length`?

Honestly I hadn't really thought that hard about that design until now, I assumed there was a good reason for the current definitions of `length`/`blocklengths`, but that seems reasonable to me and would (conceptually) simplify a lot of things from what I can tell. It makes sense that `length`/`blocklengths` should refer to the size of the actual Hilbert spaces and not the number of independent parameters (by the way, is there terminology for the dimensions that we are outputting right now besides "number of independent parameters"?). But I do think it would have nontrivial implications in `SymmetrySector.jl` since all of those lengths are referring to the number of independent parameters rather than the dimensions of the space, which one should we actually store in the `GradedUnitRange`?

Reply:
Also, it slightly breaks the abstraction we have with the data matrix of the fusion tensor, since the size of the axes won't be the same as the sizes of the actual data we are storing. In principle that's ok, but I'm not clear on how we will handle that in practice so we'll have to think that through.

Reply:
I think there is no real way to get around the size of the data matrix differing from the actual size, and this is really what you would want.
I do think that maybe the abstraction is that the datamatrix really is `Dict{SymmetrySector,Matrix}` (or maybe a vector of pairs, depending on if you want to allow duplications or not). The actual matrix really is the direct sum of `kron(identity(quantumdim(sector)), matrix)`, so the combination of just the data blocks by themselves is a bit of an ill-defined object. For example, the eigenvalues and singular values have to be duplicated, and correspondingly the norm is not simply the norm of the separate blocks. In that sense, I would kind of say that thinking of the "summed size" of the stored data matrices might hurt more then it helps, since you never really want to handle the object that is just the direct sum of these blocks instead of with the kronecker products in there.

Reply:
I see your point, but also the abstraction _does_ work for a variety of operations, for example tensor/matrix multiplication, SVD, etc. can be implemented as those operations on the block sparse/diagonal data matrix, though as you say for some of them like SVD when you are truncating you have to take into account that the singular values are repeated. So my motivations/goals would be:
1. Use the `BlockSparseMatrix` data structure for the data matrix so we can leverage block sparse SVD, matrix multiplication, etc. that we are defining in `BlockSparseArrays.jl`, and
2. Make the abstraction as explicit as possible through wrapper types like `GroupSymmetricBlockSparseMatrix`, since in the end this is just another sparse data structure with rules about what elements you store and how those map to the elements of the actual matrix.

So with that kind of `GroupSymmetricBlockSparseMatrix` design, the way I would see it is that we could achieve what you are describing in a more explicit way, by defining things like:
```julia
function Base.:*(a::GroupSymmetricBlockSparseMatrix, b::GroupSymmetricBlockSparseMatrix)
  return GroupSymmetricBlockSparseMatrix(parent(a) * parent(b), axes(a, 1), axes(b, 2))
end

function LinearAlgebra.norm(a::GroupSymmetricBlockSparseMatrix)
  # Norm of the blocks of `a` scaled by the quantum dimension
end

function LinearAlgebra.svd(a::GroupSymmetricBlockSparseMatrix)
  U, S, V = svd(parent(a))
  # Attach `axes(a)` to `U` and `V`
end

function truncated_svd(a::GroupSymmetricBlockSparseMatrix)
  # Perform `truncated_svd` taking into account the added degeneracy of the singular values
end
```
etc. Then in that way, those corresponding operations on the `FusionTensor` can just map to the same operation on the `GroupSymmetricBlockSparseMatrix` that is being stored as the data matrix.

Reply:
I like that wrapper idea. So with that change, and with further features completed for the `BlockSparseMatrix` type would a lot of code in this package become duplicate and get removed? (Just to make sure I'm following.) 

I'm glad to see this get caught about the difference between `length.(axes(ft))` and `size(ft)` since this new type represents one of the biggest changes of any storage type we've had so far, relative to the other ones. I just mean in the sense that it's still not 100% clear to me from an ITensor user level what `dim` will return on tensor indices and how `maxdim` will be interpreted in the SVD etc. I was looking forward to discussing what the best choices are there.

Reply:
> I like that wrapper idea. So with that change, and with further features completed for the `BlockSparseMatrix` type would a lot of code in this package become duplicate and get removed? (Just to make sure I'm following.)

That would be the goal, see a prototype I shared in the "symmetric_tensors" Slack channel. The current design already tries to ensure as much functionality as possible can get implemented in the `BlockSparseArrays.jl` library, this design proposal is meant to ensure even more of it can be, while taking the `size`/`axes` seriously.

> I'm glad to see this get caught about the difference between `length.(axes(ft))` and `size(ft)` since this new type represents one of the biggest changes of any storage type we've had so far, relative to the other ones. I just mean in the sense that it's still not 100% clear to me from an ITensor user level what `dim` will return on tensor indices and how `maxdim` will be interpreted in the SVD etc. I was looking forward to discussing what the best choices are there.

I think basically from the user perspective they should be defined such that they map to the equivalent definition for the abelian counterpart of the same tensor, from a generic code perspective a user should be able to switch between `U1` and `SU2` symmetry for the same system and keep the truncation parameters the same, be able to interpret the tensor sizes in the same way, etc. (obviously they won't get the exact same result since the variational space changes, but things like tolerances in the SVD, sizes of tensors, etc. should have the same interpretation). So, that is another reason that we should define things such that `length.(axes(ft)) == size(ft) == length.(axes(BlockSparseArray(ft))) == size(BlockSparseArray(ft))`.